streams/src/main/java/org/apache/kafka/streams/processor/internals/TopicPartitionMetadata.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.processor.internals;
+
+import java.nio.ByteBuffer;
+import java.util.Base64;
+import java.util.Objects;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Metadata to be committed together with TopicPartition offset
+ */
+public class TopicPartitionMetadata {
+
+    private static final Logger LOG = LoggerFactory.getLogger(TopicPartitionMetadata.class);
+
+    // visible for testing
+    static final byte LATEST_MAGIC_BYTE = 2;
+
+    private final long partitionTime;
+    private final ProcessorMetadata processorMetadata;
+
+    static TopicPartitionMetadata with(final long partitionTime, final ProcessorMetadata processorMetadata) {
+        return new TopicPartitionMetadata(partitionTime, processorMetadata);
+    }
+
+    private TopicPartitionMetadata(final long partitionTime, final ProcessorMetadata processorMetadata) {
+        Objects.requireNonNull(processorMetadata);
+        this.partitionTime = partitionTime;
+        this.processorMetadata = processorMetadata;
+    }
+
+    public long partitionTime() {
+        return partitionTime;
+    }
+
+    public ProcessorMetadata processorMetadata() {
+        return processorMetadata;
+    }
+
+    public String encode() {
+        final byte[] serializedMeta = processorMetadata.serialize();
+        // Format: MAGIC_BYTE(1) + PartitionTime(8) + processMeta
+        final ByteBuffer buffer = ByteBuffer.allocate(Byte.BYTES + Long.BYTES + serializedMeta.length);
+        buffer.put(LATEST_MAGIC_BYTE);
+        buffer.putLong(partitionTime);
+        buffer.put(serializedMeta);
+        return Base64.getEncoder().encodeToString(buffer.array());
```

Comment:
Still wondering, why we encode `processorMetadata` first as `byte[]` and convert it to `String` using Base64 here? Why not convert the `Map` to `String` directly`? The key is already String, so we would only use Base64 to convert the Long-values to String and concatenate all kv-pairs in some format (eg, CSV?)

Reply:
The reason is after converting single partitionTime to String, the length is not deterministic and we need other meta to encode it. Also I don't want to use Base64 multiple times.